enphase envoy docs quality improvement phase 2 (#36272)
home-assistant/home-assistant.github.io: source/_integrations/enphase_envoy.markdown

## Capabilities

This integration offers various entities depending on the configuration of your Enphase system. The Envoy can communicate with Enphase IQ micro-inverters, Enphase ACB and IQ batteries, Enphase Ensemble Enpower switch and load shedding relays and Enphase compatible generators.

{% note %}

- You may know referred product names under other names from the past or future.
- In this documentation, the term SN is used for a device serial-number placeholder. Entities will contain the actual serial-number of the device.
{% endnote %}

### Solar production data

All Envoy models with solar inverters, with or without installed production <abbr title="current transformers">CT</abbr>, for all firmware versions, report solar current and historical production data.

#### Aggregated production data

The Envoy device reports aggregated data for all connected micro-inverters.

> ##### Production Sensor Entities

Reviewer: ```suggestion
##### Production sensor entities
```

Author: applied

- **Envoy <abbr title="Envoy serial number">SN</abbr> Current power production**: Current aggregated inverter power production in W.
- **Envoy <abbr title="Envoy serial number">SN</abbr> Energy production last seven days**: Energy produced in previous 7 days, not including today's, in Wh, display scaled to kWh. (See known limitations [Energy Incorrect](#energy-incorrect)). This entity is not logged in statistics.
- **Envoy <abbr title="Envoy serial number">SN</abbr> Energy production today**: Energy produced since midnight in Wh, default display scaled to kWh. (See known limitations [Late reset](#late-reset), [Energy Incorrect](#energy-incorrect))

> - **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime energy production**: Lifetime energy production in Wh, default display scaled to MWh. (See known limitations [Lifetime reset](#lifetime-reset))

Reviewer: ```suggestion
- **Envoy <abbr title="Envoy serial number">SN</abbr> Energy production today**: Energy produced since midnight in Wh, default display scaled to kWh. (See known limitations [Late reset](#late-reset), [Energy Incorrect](#energy-incorrect)).
- **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime energy production**: Lifetime energy production in Wh, default display scaled to MWh. (See known limitations [Lifetime reset](#lifetime-reset)).
```

Author: implemented

<figure>
  <img src="/images/integrations/enphase_envoy/enphase_envoy_solar_production.png" alt="Envoy device">
  <figcaption>Envoy device with solar production entities.</figcaption>
</figure>

When used with [multiphase CT phase data](#ct-aggregate-and-phase-data), disabled phase entities are available as well.

#### Individual micro-inverter production data

The Envoy reports individual micro-inverter device production data. SN is the micro-inverter serial-number.

> ##### Sensor Entities

Reviewer: ```suggestion
##### Sensor entities
```

Author: applied

- The storage CT measures the energy exchange between the battery storage and the switchboard.

<figure>
  <img src="/images/integrations/enphase_envoy/enphase_envoy_ct_model.png" alt="ct model">
  <figcaption>Envoy current transformers.</figcaption>
</figure>

{% note %}
This in no way represents a configuration direction, as actual configuration is driven by local rules, installer designs and Enphase installation guidelines. Variations may exist, based on specific needs or rules. This merely describes a simplified view and naming conventions used in this documentation to clarify integration operation. Refer to [Enphase documentation](https://enphase.com/installers/resources/documentation/communication) for more information.
{% endnote %}

When an Envoy Metered is equipped with a production CT, the CT data will be used to provide the [aggregated solar production data](#aggregated-production-data). Likewise, the installed consumption CT will be the source for the [house consumption data](#house-consumption-data).

Either a net-consumption or a total-consumption CT is installed. The Envoy will calculate the data for the other one.

#### CT Aggregate and phase data

> When using <abbr title="current transformers">[CT](#current-transformers)</abbr> in multiphase enabled configurations, both aggregated and individual phase data is available. If only 1 phase is configured, phase data is the same as the aggregated data and no phase entities are created. If more than 1 phase are configured and measured, individual, disabled, phase entities are available, enable them as needed.

Reviewer: I'm not sure I understand this sentence: 

> " If more than 1 phase are configured and measured, individual, disabled, phase entities are available, enable them as needed."

Do you mean something like this?
If more than 1 phase is being measured: If the phases are disabled individually, enable them as needed.

Author: Too much packed in one sentence.

If more then 1 phase is available, disabled phase entities for each phase are created, enable them as needed.

Author: Changed to: If more than 1 phase is configured, disabled phase entities for each phase are created, enable them as needed.

> Phase entity names are the names used for the aggregated entities, post fixed with the phase name. Phase names are **L1**, **L2**, **L3**. For example, once enabled, [**lifetime energy production**](#solar-production-data) on phase 3 is available as **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime energy production L3**.

Reviewer: ```suggestion
Phase entity names are the names used for the aggregated entities, with the phase name in the post-fix. Phase names are **L1**, **L2**, **L3**. For example, once enabled, [**lifetime energy production**](#solar-production-data) on phase 3 is available as **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime energy production L3**.
```

Author: applied

- **Envoy <abbr title="Envoy serial number">SN</abbr> Voltage production CT**: Voltage in V. (see limitations [Summed voltage](#summed-voltage))
- **Envoy <abbr title="Envoy serial number">SN</abbr> Production CT current**: Current in A.
- **Envoy <abbr title="Envoy serial number">SN</abbr> Powerfactor production CT**: Powerfactor, ratio of active to apparent power.
- **Envoy <abbr title="Envoy serial number">SN</abbr> Metering status production CT**: Status of the metering process: `normal`, `not-metering`, `check-wiring`.
- **Envoy <abbr title="Envoy serial number">SN</abbr> Meter status flags active production CT**: Count of CT status flags active. See troubleshooting [CT Active flag count](#ct-active-flag-count) when non-zero.

##### Net-consumption CT sensor entities

- **Envoy <abbr title="Envoy serial number">SN</abbr> Frequency net consumption CT**: Frequency in Hz .
- **Envoy <abbr title="Envoy serial number">SN</abbr> Voltage net consumption CT**: Voltage in V. (see limitations [Summed voltage](#summed-voltage)
- **Envoy <abbr title="Envoy serial number">SN</abbr> net consumption CT current**: Current in A.

> - **Envoy <abbr title="Envoy serial number">SN</abbr> Powerfactor net consumption CT**: Powerfactor, ratio of active to apparent power .

Reviewer: ```suggestion
- **Envoy <abbr title="Envoy serial number">SN</abbr> Power factor net consumption CT**: Power factor, ratio of active to apparent power.
```
I'm unsure about the term powerfactor. If [this power factor](https://www.merriam-webster.com/dictionary/power%20factor) is meant, it would be spelled in 2 words. But maybe there is a reason to spell it as one word?

Author: Yes it's the power factor you refer too. But in this case it's the literal name the entity is given in the code. User will see this on the screen (at least when using english as language). They probbly won't notice if we add a space here, but I rather opt for 'as on screen'.

- **Envoy <abbr title="Envoy serial number">SN</abbr> Metering status net consumption CT**: Status of the metering process: `normal`, `not-metering`, `check-wiring`.
- **Envoy <abbr title="Envoy serial number">SN</abbr> Meter status flags active net consumption CT**: Count of CT status flags active. See troubleshooting [CT Active flag count](#ct-active-flag-count) when non-zero.

##### Storage CT sensor entities

> - **Envoy <abbr title="Envoy serial number">SN</abbr> Frequency storage CT**: Frequency in Hz .

Reviewer: ```suggestion
- **Envoy <abbr title="Envoy serial number">SN</abbr> Frequency storage CT**: Frequency in Hz.
```

- **Envoy <abbr title="Envoy serial number">SN</abbr> Voltage storage CT**: Voltage in V. (see limitations [Summed voltage](#summed-voltage)
- **Envoy <abbr title="Envoy serial number">SN</abbr> storage CT current**: Current in A.

> - **Envoy <abbr title="Envoy serial number">SN</abbr> Powerfactor storage CT**: Powerfactor, ratio of active to apparent power .

Reviewer: ```suggestion
- **Envoy <abbr title="Envoy serial number">SN</abbr> Power factor storage CT**: Power factor, ratio of active to apparent power .
```

- **Envoy <abbr title="Envoy serial number">SN</abbr> balanced net power consumption**: Current power exchange from (positive) / to (negative) the grid in W, default display in kW.
  (This is the same value as [Envoy <abbr title="Envoy serial number">SN</abbr> Current net power consumption](#grid-sensor-entities) when using a net-consumption CT.)
- **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime balanced net energy consumption**: Lifetime energy balance (difference) of imported and exported grid energy in Wh, default display in kWh.

When used with [multiphase CT phase data](#ct-aggregate-and-phase-data), disabled phase entities are available as well.

### Battery Storage data

Enphase battery systems of multiple generations and configurations can provide integration entity data.

- **AC-Batteries**: first generation battery setup, no longer in production.
- **IQ Batteries**: current Enphase battery models.

The batteries can be implemented in stand-alone configuration or as part of Enphase Ensemble systems with the Enpower/IQ System Controller.  

#### IQ battery data

> ##### Aggregated IQ battery Sensor entities

Reviewer: ```suggestion
##### Aggregated IQ battery sensor entities
```

- **Envoy <abbr title="Envoy serial number">SN</abbr> Reserve battery level**: Configured aggregated IQ Battery backup state of charge in %
- **Envoy <abbr title="Envoy serial number">SN</abbr> Reserve battery energy**: Configured aggregated IQ battery backup energy content in Wh

If a [storage <abbr title="current transformers">CT</abbr>](#storage-ct-sensor-entities) is installed:

- **Envoy <abbr title="Envoy serial number">SN</abbr> Current battery discharge**: Current power in/out of the battery in W.
- **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime battery energy discharged**: Lifetime energy discharged from the battery in Wh, default display format MWh.
- **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime battery energy charged**: Lifetime energy charged in the battery in Wh, default display format MWh.

<figure id="aggregated-iq-battery-data">
  <img src="/images/integrations/enphase_envoy/enphase_envoy_aggr_iq_battery.png" alt="aggregated iq battery">
  <figcaption>Envoy Aggregated IQ battery Sensor entities with no <a href="#current-transformers">storage CT</a> nor <a href="#enpower-data">Enpower</a> installed.</figcaption>
</figure>

{% note %}

> If an Enpower device is installed, then **Charge from grid**, **Reserve battery level** and **Storage mode** are available as [Enpower device entities](#enpower-binary-sensor-entities) instead of Envoy device entities.

Reviewer: ```suggestion
If an Enpower device is installed, then **Charge from grid**, **Reserve battery level**, and **Storage mode** are available as [Enpower device entities](#enpower-binary-sensor-entities) instead of Envoy device entities.
```

Author: applied

{% endnote %}

#### Individual IQ battery data

For each IQ Battery, an Encharge device is created, linked to the Envoy parent device. Each encharge devices offers individual battery sensors.

> ##### Individual IQ battery Sensor entities

Reviewer: ```suggestion
##### Individual IQ battery sensor entities
```

- **Encharge <abbr title="Encharge serial number">SN</abbr> Battery**: Current state of charge of the battery in %
- **Encharge <abbr title="Encharge serial number">SN</abbr> Power**: Current power in W
- **Encharge <abbr title="Encharge serial number">SN</abbr> Apparent Power**: Current apparent power in VA
- **Encharge <abbr title="Encharge serial number">SN</abbr> Temperature**: Current temperature in degrees C or F, based on your localization.
- **Encharge <abbr title="Encharge serial number">SN</abbr> Last reported**: Time when Envoy received last update from the battery.

> ##### Individual IQ battery Binary sensor entities

Reviewer: ```suggestion
##### Individual IQ battery binary sensor entities
```

- **Encharge <abbr title="Encharge serial number">SN</abbr> Communicating**: Communication status of encharge battery, Connected / Disconnected. This is a diagnostics entity.
- **Encharge <abbr title="Encharge serial number">SN</abbr> DC Switch**: DC Switched off status of encharge battery, On / Off. This is a diagnostics entity.

<figure>
  <img src="/images/integrations/enphase_envoy/enphase_envoy_encharge.png" alt="iq battery">

>   <figcaption>Envoy Encharge IQ battery Sensor entities.</figcaption>

Reviewer: ```suggestion
  <figcaption>Envoy Encharge IQ battery sensor entities.</figcaption>
```

</figure>

> #### AC-Battery data

Reviewer: ```suggestion
#### AC-battery data
```

> No individual AC-battery data is available, only aggregated AC-Battery data for all AC-batteries.

Reviewer: ```suggestion
No individual AC-battery data is available, only aggregated AC-battery data for all AC-batteries.
```

> ##### AC-battery Sensor entities

Reviewer: ```suggestion
##### AC-battery sensor entities
```

- **ACB <abbr title="Envoy serial number">SN</abbr> Battery**: Current AC-battery state of charge in %
- **ACB <abbr title="Envoy serial number">SN</abbr> Battery state**: AC-battery state: charging, idle, discharging
- **ACB <abbr title="Envoy serial number">SN</abbr> Power**: Current AC-battery power in W
- **Envoy <abbr title="Envoy serial number">SN</abbr> Available ACB battery energy**: Current AC-battery energy content in Wh

<figure>
  <img src="/images/integrations/enphase_envoy/enphase_envoy_acb_battery.png" alt="acb battery">

>   <figcaption>Envoy ACB battery Sensor entities.</figcaption>

Reviewer: ```suggestion
  <figcaption>Envoy ACB battery sensor entities.</figcaption>
```

</figure>

> ##### Aggregated IQ and AC battery Sensor entities

Reviewer: ```suggestion
##### Aggregated IQ and AC battery sensor entities
```

> If both IQ and AC batteries are used, aggregated battery data for all installed IQ Batteries and AC-Batteries is available.

Reviewer: ```suggestion
If both IQ and AC batteries are used, aggregated battery data for all installed IQ batteries and AC batteries is available.
```

- **Envoy <abbr title="Envoy serial number">SN</abbr> Aggregated Battery SOC**: Overall aggregated battery state of charge in %
- **Envoy <abbr title="Envoy serial number">SN</abbr> Aggregated Available battery energy**: Overall aggregated battery energy content in Wh
- **Envoy <abbr title="Envoy serial number">SN</abbr> Aggregated Battery capacity**: Overall aggregated maximum battery energy content in Wh

> The below figure shows the 3 Aggregated entities along with the [AC_Battery energy](#ac-battery-sensor-entities) and 3 of the [IQ Battery aggregate](#aggregated-iq-battery-sensor-entities) values.

Reviewer: ```suggestion
The below figure shows the 3 aggregated entities along with the [AC_Battery energy](#ac-battery-sensor-entities) and 3 of the [IQ Battery aggregate](#aggregated-iq-battery-sensor-entities) values.
```

<figure>
  <img src="/images/integrations/enphase_envoy/enphase_envoy_aggr_acb_iq_battery.png" alt="aggregated acb iq battery">

>   <figcaption>Envoy Aggregated ACB and IQ battery Sensor entities.</figcaption>

Reviewer: ```suggestion
  <figcaption>Envoy aggregated ACB and IQ battery sensor entities.</figcaption>
```

</figure>

### Enpower data

Data for an installed Enphase Enpower off-grid enabling switch is available in various entities for an Enpower device.

> #### Enpower Binary sensor entities

Reviewer: ```suggestion
#### Enpower binary sensor entities
```

- **Enpower <abbr title="Enpower serial number">SN</abbr> Grid status**: status of the grid.

> #### Enpower Number entities

Reviewer: ```suggestion
#### Enpower number entities
```

Author: applied

- **Enpower <abbr title="Enpower serial number">SN</abbr> Reserve battery level**: reserve battery level to maintain for outages in %. Changing the value, on the UI or in an [action](#action-numberset_value), will update the setting in the Envoy. Also see limitations, [No battery controls](#no-battery-controls).

> #### Enpower Select entities

Reviewer: ```suggestion
#### Enpower select entities
```

Author: Applied

- **Enpower <abbr title="Enpower serial number">SN</abbr> Storage mode**: Current configured storage mode, `Full backup`, `Self consumption`, `Savings mode`. Changing the selection, in the UI or in an [action](#action-selectselect), will update the setting in the Envoy. Also see limitations, [No battery controls](#no-battery-controls).

> #### Enpower Sensor entities

Reviewer: ```suggestion
#### Enpower sensor entities
```

Author: Appplied

- **Enpower <abbr title="Enpower serial number">SN</abbr> Temperature**: Current temperature in degrees C or F, based on your localization.
- **Enpower <abbr title="Enpower serial number">SN</abbr> Last reported**: Time when Envoy received last update from the enpower device.
- **Enpower <abbr title="Encharge serial number">SN</abbr> Communicating**: Communication status of enpower switch, Connected / Disconnected. This is a diagnostics entity.

> #### Enpower Switch entities

Reviewer: ```suggestion
#### Enpower switch entities
```

Author: Apllied

- **Enpower <abbr title="Enpower serial number">SN</abbr> Charge from grid**: Allow or disallow charging Encharge/IQ batteries from grid when a charge schedule is active. Changing the switch, in the UI or in an [action](#action-switchturn_onswitchturn_offswitchtoggle), will update the setting in the Envoy. If no charge schedule is active, changing the setting will not have an immediate effect. Also see limitations, [No battery controls](#no-battery-controls).
- **Enpower <abbr title="Enpower serial number">SN</abbr> Grid enabled**: Enable or disable grid connection. Note that the Enpower has a slight delay built-in between receiving these commands and actually switching the system on or off grid.

<figure>
  <img src="/images/integrations/enphase_envoy/enphase_envoy_enpower.png" alt="envoy enpower">
  <figcaption>Envoy Enpower entities.</figcaption>
</figure>

{% note %}

> If no Enpower is installed, then **Charge from grid**, **Reserve battery level** and **Storage mode** are available as [Envoy device entities](#aggregated-iq-battery-data) instead of Enpower device entities.

Reviewer: ```suggestion
If no Enpower is installed, then **Charge from grid**, **Reserve battery level**, and **Storage mode** are available as [Envoy device entities](#aggregated-iq-battery-data) instead of Enpower device entities.
```

{% endnote %}

### Enpower load shedding relays

With the Enpower/IQ System Controller installed, entities for the state and control of the Enpower's four load-shedding relays are available. In battery installations without load-shedding and off-grid functionality, used in many EU countries, these sensors and switches are not available.

> The Enphase Enpower has 4 load shedding relays, dry-contacts, that can be used to control non-essential loads in your home. These have two main modes of operation:

Reviewer: ```suggestion
The Enphase Enpower has 4 load shedding relays (volt-free contacts) that can be used to control non-essential loads in your home. These have two main modes of operation:
```
replaced dry contact by volt-free contact because I feel it is easier to understand. feel free to ignore this comment if it is incorrect.

Author: merged the 2: (volt-free contacts, often referred to as dry-contacts) as dry-contacts shows later in the doc.

- **Standard**: When the mode entity is set to standard, you can simply set the state of the relay to be powered or not powered for each mode of operation: on grid, off grid, and on generator.

> - **Battery level**: When the relay mode is set to battery level, the relays will turn on and off based on the remaining battery level of your Encharge IQ batteries. Two number entities are available to control the cutoff and restore levels for the relays. When the battery level drops below the cutoff level, the relays will turn off. When the battery level rises above the restore level, the relays will turn back on.

Reviewer: what happens when the relay turn on or off?

Author: updated next sentences to:   ...the relays will turn off, removing power from attached loads. When the battery level rises above the restore level, the relays will turn back on and restore power to attached loads.

#### Dry-contact number entities

- **<abbr title="dry_contacts.load_name">LOAD_NAME</abbr> cutoff battery level**: battery level below which relay will turn off when in `Battery level` mode.
- **<abbr title="dry_contacts.load_name">LOAD_NAME</abbr> restore battery level**: battery level above which the relay will turn back on when in `Battery level` mode.

#### Dry-contact select entities

- **<abbr title="dry_contacts.load_name">LOAD_NAME</abbr> mode**: dry-contact operational mode: `standard` or `battery`
- **<abbr title="dry_contacts.load_name">LOAD_NAME</abbr> Grid action**: dry-contact on grid action: `Powered`, `Not powered`, `Follow schedule`, `None`
- **<abbr title="dry_contacts.load_name">LOAD_NAME</abbr> Micro-grid action**: dry-contact on micro-grid action: `Powered`, `Not powered`, `Follow schedule`, `None`
- **<abbr title="dry_contacts.load_name">LOAD_NAME</abbr> Generator action**: dry-contact on generator action:  `Powered`, `Not powered`, `Follow schedule`, `None`

<figure>
  <img src="/images/integrations/enphase_envoy/enphase_envoy_dry_contact.png" alt="envoy dry-contact">
  <figcaption>Envoy Enpower dry-contact entities.</figcaption>
</figure>

> ## Data polling Interval

Reviewer: ```suggestion
## Data polling interval
```

> All data is collection in one coordinated collection cycle, minimizing the number of requests to the Envoy. The local REST API of the Envoy is used. Only when the 1-year valid token is to expire, 1 month before due data, a new token is requested from the Enphase Enlighten web-site.

Reviewer: ```suggestion
All data is collected in one coordinated collection cycle. For example, three different values are not pulled in three different requests but bundled in one single request. This method minimizes the number of requests to the Envoy. The local REST API of the Envoy is used. Only when the 1-year valid token is to expire, 1 month before due data, a new token is requested from the Enphase Enlighten website.
```
I struggled a bit with "All data is collected in one coordinated collection cycle. " Maybe we can add an example? Please correct my suggestion if I misunderstood the meaning.

Author: Changed to: All data is collected in one coordinated collection cycle and sourced from a limited set of endpoints on the Envoy. For example, three different values sourced from the same endpoint, are not pulled in three different requests, but provided from the same single request. This method minimizes the number of requests to the Envoy.

> All data is collected in one coordinated collection cycle and sourced from a limited set of endpoints on the Envoy. For example, three different values sourced from the same endpoint, are not pulled in three different requests, but provided from the same single request. This method minimizes the number of requests to the Envoy. The local REST API of the Envoy is used. Only when the 1-year valid token is to expire, 1 month before due data, a new token is requested from the Enphase Enlighten website.

Reviewer: ```suggestion
All data is collected in one coordinated collection cycle and sourced from a limited set of endpoints on the Envoy. For example, three different values sourced from the same endpoint are not pulled in three different requests but provided from the same single request. This method minimizes the number of requests to the Envoy. The local REST API of the Envoy is used. Only when the 1-year valid token is to expire, 1 month before due data, a new token is requested from the Enphase Enlighten website.
```

Envoy installations without installed <abbr title="current transformers">CT</abbr>, collect individual solar inverter data every 5 minutes. This collection does not occur for each inverter at the same time in the 5-minute period. Shortening the collection interval will at best show updates for individual inverters quicker, but not yield more granular data.

With installed <abbr title="current transformers">CT</abbr>, data granularity increases and shortening the collection interval can provide more details. The Envoy, however, has no unlimited resources and shortening the collection interval may result in dropped connections, Envoy freeze or restarts. It will require some step-wise tuning for each individual situation.

## Energy dashboard

This integration provides several entities suitable for the energy dashboard.

### Solar panels

For **Solar production**, use the **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime energy production** entity. Overall this has proven to be more stable as the daily value reported by the Envoy. (See known limitations [Late reset](#late-reset), [Energy Incorrect](#energy-incorrect)).

> ### Electricity Grid

Reviewer: ```suggestion
### Electricity grid
```

> Wether data to use with the electricity grid is available, depends on installed <abbr title="current transformers">CT</abbr>, if any. Also see Limitations, [Balancing grid meter](#balancing-grid-meter).

Reviewer: ```suggestion
Whether there is data available to use with the electricity grid depends on the installed <abbr title="current transformers">CT</abbr>, if any. Also, see Limitations, [Balancing grid meter](#balancing-grid-meter).
```
Not sure if I understand this correctly. What data are we talking about there? energy consumption?

Author: applied

> #### Electricity Grid with net-consumption CT

Reviewer: ```suggestion
#### Electricity grid with net-consumption CT
```

With a [net-consumption CT](#grid-sensor-entities) installed, both grid consumption and return to grid data is available.

- For **Grid consumption**, use the **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime net energy consumption** entity.
- For **Return to grid**, use the **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime net energy production** entity.

> #### Electricity Grid with total-consumption CT

Reviewer: ```suggestion
#### Electricity grid with total-consumption CT
```

{% raw %}

```yaml

- trigger:
    - platform: state
      entity_id: sensor.envoy_sn_lifetime_balanced_net_energy_consumption
  
  sensor:
    - name: "Grid import"
      unique_id: calculated_envoy_grid_import
      unit_of_measurement: "Wh" 
      state: "{{ this.state | int(0) + ([0, (trigger.to_state.state | int(0) - trigger.from_state.state | int(0))] | max }}" 
      device_class: energy 
      state_class: total_increasing 
    - name: "Grid export"
      unique_id: calculated_envoy_grid_export
      unit_of_measurement: "Wh" 
      state: "{{ this.state | int(0) - ([0, (trigger.to_state.state | int(0) - trigger.from_state.state | int(0))] | min }}" 
      device_class: energy 
      state_class: total_increasing
```

{% endraw %}

The above example does not address handling `unavailable` or `unknown` states, value changes over Home Assistant outages nor changing UOM to a preferred one. Examples for these exist in various community topics.

{% tip %}

> Alternatively, this can be done splitting  **Envoy <abbr title="Envoy serial number">SN</abbr> balanced net power consumption** into power import and export and two Riemann sum integral helpers to calculate energy from the power values.

Reviewer: ```suggestion
Alternatively, this can be done by splitting the **Envoy <abbr title="Envoy serial number">SN</abbr> balanced net power consumption** into power import and export and two Riemann sum integral helpers to calculate energy from the power values.
```

{% endtip %}

{% enddetails %}

### Home battery storage

> Wether data to use with the Home battery storage is available, depends on installed storage <abbr title="current transformers">CT</abbr>, if any.

Reviewer: ```suggestion
Whether there is data available to use with the electricity grid depends on the installed storage <abbr title="current transformers">CT</abbr>, if any.
```

Author: Applied.

#### Home battery storage with storage CT

With a [storage CT](#aggregated-iq-battery-sensor-entities) installed, data for both Energy coming out and going into the battery is available.

- For **Energy going into the battery**, use the **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime battery energy charged** entity.

> - For **Energy coming out off the battery**, use the **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime battery energy discharged** entity.

Reviewer: ```suggestion
- For **Energy coming out of the battery**, use the **Envoy <abbr title="Envoy serial number">SN</abbr> Lifetime battery energy discharged** entity.
```

The concept is to first sum all battery Power values using a combine state helper. Then track value changes of the summed value entity, add positive changes to a battery_charge power entity and add negative changes to a battery_discharge power entity.

{% raw %}

```yaml

- trigger:
    - platform: state
      entity_id: sensor.envoy_sn_summed_battery_power_entity
  
  sensor:
    - name: "Battery charge power"
      unique_id: calculated_envoy_battery_charge_power
      unit_of_measurement: "W" 
      state: "{{ this.state | int(0) + ([0, (trigger.to_state.state | int(0) - trigger.from_state.state | int(0))] | max }}" 
      device_class: power 
      state_class: measurement
    - name: "Battery discharge power"
      unique_id: calculated_envoy_battery_discharge_power
      unit_of_measurement: "W" 
      state: "{{ this.state | int(0) - ([0, (trigger.to_state.state | int(0) - trigger.from_state.state | int(0))] | min }}" 
      device_class: power 
      state_class: measurement
```

{% endraw %}

The above example does not address handling `unavailable` or `unknown` states, value changes over Home Assistant outages nor conversion losses.

{% enddetails %}

> ##### Home battery storage data using Available battery energy

Reviewer: ```suggestion
##### Home battery storage data on the available battery energy
```

Author: Applied

## Know issues and limitations

### Firmware changes

[Envoy firmware](https://enphase.com/installers/resources/documentation/communication?&f[0]=document%3A217) versions come with changing behavior, features and issues. Firmware is pushed to the Envoy by Enphase, while 'not always communicated in detail upfront'. This may result in sudden changes in the Envoy behavior and is always accompanied by an outage while Envoy is being updated.

### No battery controls

As of Envoy firmware  8.2.4225, the Envoy no longer supports setting battery modes, enabling/disabling charging from the grid, or changing reserve battery level through the local REST API used by HA. Until a resolution is found, you will need to use the Enphase APP to control these.

> 

Author: I's already in the Know issues and limitations paragraph

The Envoy Metered in multiphase setup, sums the voltages of the phases measured on the CT for the aggregated data. This may be valid for split-phase, but for 3-phase systems, use the individual phases rather than the summed value.

### Balancing grid meter

In multiphase installations with batteries, in countries with phase-balancing grid meters, the battery will export to the grid on one phase the amount it lacks on another phase. This other phase pulls the missing amount from the grid, as if it is using the grid as a 'transport' between phases. Since the grid meter will balance the amount imported and exported on the two phases, the net result is zero. The Envoy multiphase net-consumption CTs, however, will report the amounts on both phases, resulting in too high export on one and too high import on the other. One may consider using the `lifetime balanced net energy consumption` which is the sum of grid import and export to eliminate this effect. This would require some templating to split these values into import and export values. Alternatively, use the `current net power consumption` or `balanced net power consumption` with a Riemann integral sum helper.

## Troubleshooting

> ### Single network

Reviewer: ```suggestion
### Periodic network connection issues
```
I think the title of a troubleshooting item should describe the symptom people are noticing so that it can be found more easily. Feel free to rephrase if you can think of a better description.

Author: Excellent proposal! Applied

> The Envoy should not be both on your local LAN (Ethernet) and local Wi-Fi simultaneously. Having both connections active can cause auto-discovery to alternate between interfaces, resulting in connection interruptions approximately every 30 minutes. If you experience these periodic connection issues, ensure the Envoy is connected using only one interface (either Ethernet OR Wi-Fi, not both).

Reviewer: ```suggestion
If you experience periodic connection issues, ensure the Envoy is connected to only one interface (either Ethernet OR Wi-Fi, not both). The Envoy should not be both on your local LAN (Ethernet) and local Wi-Fi simultaneously. Having both connections active can cause auto-discovery to alternate between interfaces, resulting in connection interruptions approximately every 30 minutes.
```
rearranged the text a bit to start with a description of the symptom, followed by steps to mitigate the issue. And only then go into an explanation.

Author: Great improvement, thanks. Applied

> ### CT Active flag count

Reviewer: ```suggestion
### CT Active flag count is non-zero
```

Author: Applied

> This value is the count of CT status flags that are raised. In normal state, the value is zero. If the value is non-zero, consult the [diagnostic](#diagnostics) report of the Envoy and look for `raw_data` - `/ivp/meters` - `statusFlags` for set flags, one or more from  (`production-imbalance` | `negative-production` | `power-on-unused-phase` | `negative-total-consumption`). Their somewhat descriptive names may be an indication of installation issues.

Reviewer: ```suggestion
The **CT active flag count** value shows the number of CT status flags that are raised. In a normal state, the value of **CT active flag count** is zero. If the value is non-zero, consult the [diagnostic](#diagnostics) report of the Envoy and look for `raw_data` - `/ivp/meters` - `statusFlags` for set flags, one or more from  (`production-imbalance` | `negative-production` | `power-on-unused-phase` | `negative-total-consumption`). Their somewhat descriptive names may be an indication of installation issues.
```

Author: Applied, again excellent improvement. Thanks so much!
